make automatic validation optional

ernie58 opened the pull request:
make automatic validation optional via CRUD-options or via setting in model definition

@bajtos , @superkhau What should I do with the validateUpsert-option? Which one needs precedence? Now, when explicitly set to false, it has precedence on the new flags, which makes sense IMO

Close #744

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

ernie58 commented:
pull request for #744

bajtos commented on lib/dao.js:
```diff
@@ -257,6 +257,16 @@ DataAccessObject.create = function (data, options, cb) {
     data = obj.toObject(true);
 
+    //options has precedence on model-setting
+    if(options.disableValidation === true){
```
Coding style:

```js
if (options.disableValidation === true) {
```

Please apply this style to the rest of your changes too.

ernie58 replied:
I thought I was doing this? Could you tell me where I don't use this style? Thx

bajtos replied:
Check the spaces - add a space after `if` and another in between `) {`.

bajtos commented:
I am proposing to rename `options.disableValidation` to `options.validation` and `disableAutomaticValidation` to `automaticValidation`. [Double negative](https://en.wikipedia.org/wiki/Double_negative) like `disableValidation:false` tends to be confusing for readers.

bajtos commented:
> What should I do with the validateUpsert-option? Which one needs precedence? Now, when explicitly set to false, it has precedence on the new flags, which makes sense IMO.

IMO, it's ok for `validateUpsert` to take precedence over `automaticValidation`. However, `options.validate` should have always the highest precedence.

bajtos commented:
@raymondfeng would you like to take a look too?

bajtos commented on test/optional-validation.test.js:
```diff
@@ -0,0 +1,504 @@
+
+  describe('save with validation errors and options.disableValidation, no model setting', function () {
+
+    it('should throw validation-errors on create(data, {disableValidation: false}, cb) with invalid data', function (done) {
```
Please shorten test case names, e.g. `it('should throw on create when options.disableValidation=false`)`

bajtos commented on test/optional-validation.test.js:
```diff
@@ -0,0 +1,504 @@
+    });
+  });
+
+  describe('save with validation errors and options.disableValidation, no model setting', function () {
```
This description is incorrect, one of the tests are actually passing a valid instance:

```
should NOT throw validation-errors on create(data, {disableValidation: true}, cb) with valid dat
```

bajtos commented:
@slnode ok to test

ernie58 commented:
I will take out the double negation, and change and shorten the descriptions in the test cases

ernie58 commented:
@bajtos I'm confused about the `options.validate` - flag. I thought it was set by `validateUpsert`, but in your comment you state that `options.validate`has precedence on `validateUpsert`. So it's used elsewhere then? Why can't I use directly `options.validate` instead of creating another option `options.validation`

slnode commented:
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

ernie58 commented:
Sorry, closed it by accident

bajtos commented:
> I'm confused about the options.validate - flag. I thought it was set by validateUpsert, but in your comment you state that options.validate has precedence on validateUpsert. So it's used elsewhere then? Why can't I use directly options.validate instead of creating another option options.validation

Actualy, AFAICT, `prototype.save` already supports `options.validate` flag - see [here](https://github.com/strongloop/loopback-datasource-juggler/blob/5ca9c47f440ab999c4dfe4a323a3d72991fd9c9f/lib/dao.js#L1878-L1880) and [here](https://github.com/strongloop/loopback-datasource-juggler/blob/5ca9c47f440ab999c4dfe4a323a3d72991fd9c9f/lib/dao.js#L1902-L1905). Which is yet another argument why to use this name instead of adding `options.disableValidation`.

Yes, you can and should use `options.validate` directly.

I believe `options.validate` is unrelated to model-level setting `validateUpsert` at the moment. The changes you are proposing will change that and I think that any option passed to a method call should always overwrite model-level settings.

ernie58 commented:
@bajtos I've added a commit, could you check again? Thx!

ernie58 commented:
@bajtos did you have a chance to look into the changes?

bajtos commented on lib/dao.js:
```diff
@@ -1899,7 +1918,6 @@ DataAccessObject.prototype.save = function (options, cb) {
     Model.applyProperties(data, inst);
     inst.setAttributes(data);
 
-    // validate first
```
I don't see why this comment is no longer relevant? Would you mind putting it back again?

bajtos commented on test/optional-validation.test.js:
```diff
@@ -0,0 +1,504 @@
+      User.validatesPresenceOf('name');
+
+      User.create({name: null}, options, function (e, d) {
+        should.not.exist(e);
```
Please use `if (err) return done(err)` - it gives more helpful stack traces on error. Also please apply the same changes to other places where you call `should.not.exist(e)`.

bajtos commented on lib/dao.js:
```diff
@@ -1876,8 +1886,17 @@ DataAccessObject.prototype.save = function (options, cb) {
 
   if (options.validate === undefined) {
-    options.validate = true;
+    if (Model.settings.validateUpsert === undefined) {
```
I believe `PersistedModel.prototype.save` is not affected by `validateUpsert` setting. I think this block of code should belong to `PersistedModel.prototype.updateAttributes`, which was not changed to support `options.validate` yet?

bajtos commented:
AFAICT, this patch adds the new option/setting only to `create()` and `prototype.save()` methods. How about other CRUD methods like `updateOrCreate()`, `findOrCreate()`, `prototype.updateAttributes()`?

I don't mind landing a change that adds `options.validate` to a subset of CRUD methods.

However, I am worried that if we land model-level `settings.automaticValidation` that supports only a subset of CRUD methods, then users will be confused.

@ernie58 thoughts?

bajtos commented on test/optional-validation.test.js:
```diff
@@ -0,0 +1,504 @@
+
+    it('should throw on create with validate:true with invalid data', function (done) {
+      var options = {validate: true};
+      User.validatesPresenceOf('name');
```
Please move this line to `beforeEach` and/or top-level model setup code, so that it is not duplicated in every test. Unless I missed something?

bajtos commented:
BTW I appreciate the extensive test coverage! 

I'm wondering how to refactor the code to reduce duplication. For example:

```js
    it('should throw on create with validate:true with invalid data', function (done) {
      var options = {validate: true};
      User.create(INVALID_DATA, options, expectValidationError(done));
    });

    it('should NOT throw on create with validate:true with valid data', function (done) {
      var options = {validate: true};
      User.create(VALID_DATA, options, expectSuccess(done));
    });

   it('should NOT throw on create with validate:false with valid data', function (done) {
      var options = {validate: false};
      User.create(VALID_DATA, options, expectSuccess(done));
    });

   // etc ...

function expectSuccess(done) {
  return function(err, result) {
    // make the assertions
  };
}
```

ernie58 commented:
@bajtos I thought those extra methods used save internally, because in #744 you only mentioned create and save. I'll add the other methods, that will be better, it's not a big effort

ernie58 commented:
@bajtos I agree on the code duplication, i'll look into it. It started small but with the double negation and the model setting, it expanded quick. I'll clean it up

bajtos commented:
>  I thought those extra methods used save internally, because in #744 you only mentioned create and save. I'll add the other methods, that will be better, it's not a big effort

Sorry for the confusion. It's great to hear that you don't mind adding support to other methods too! :+1: 

> I agree on the code duplication, i'll look into it. It started small but with the double negation and the model setting, it expanded quick. I'll clean it up

Awesome :)

ernie58 commented:
@bajtos I refactored the testing and added the optional validation to `updateAttributes`, `updateOrCreate` and `findOrCreate`. I still have some question, but I'll try to post them inline

ernie58 commented on lib/dao.js:
```diff
@@ -457,12 +467,27 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
           inst.isValid(function(valid) {
             if (!valid) {
-              if (Model.settings.validateUpsert) {
+              if (doValidate) { //backwards compatibility with validateUpsert:undefined
```
I've used the some logic as validateUpsert. When options and settings are undefined I default to false instead of true for automatic validation, only in updateOrCreate, to stay backwards compatible.

I also use a local variable `doValidate` instead of setting  `options.validate`, because some test required the options to stay the same during the lifecycle. I did that in `save`, but the tests completed successfully even when changing the `options`. Should it be changed there as well?

bajtos replied:
> I've used the some logic as validateUpsert. When options and settings are undefined I default to false instead of true for automatic validation, only in updateOrCreate, to stay backwards compatible.

Great, that's the right thing to do here :+1:

bajtos replied:
> I also use a local variable doValidate instead of setting options.validate, because some test required the options to stay the same during the lifecycle. I did that in save, but the tests completed successfully even when changing the options. Should it be changed there as well?

It's better to treat input arguments as immutable and use a local variable (e.g. `doValidate`) everywhere.

ernie58 commented on lib/dao.js:
```diff
@@ -2426,7 +2470,19 @@ DataAccessObject.prototype.updateAttributes = function updateAttributes(data, op
 
     inst.isValid(function (valid) {
-      if (!valid) {
+
```
At first I did not call the `isValid`-function, when no validation was needed. But then some test concerning the lifecycle hooks failed, because the `beforeValidation` and `afterValidation` hooks where missing. I was not sure if I could modify those, so instead I always validate, and depending on the settings and options I throw a ValidationError or I don't

bajtos replied:
> But then some tests concerning the lifecycle hooks failed, because the beforeValidation and afterValidation hooks where missing. 

Hmm, that's interesting and and rather unexpected. Could you please provide more details on the test failures? I am surprised that an existing test fails when this change should be fully backwards compatible.

ernie58 replied:
guess i did something wrong here, changed it again, but now it works. isValid is now only called when needed

ernie58 commented:
@bajtos I checked the docs for other CRUD functions, I don't know if any of these apply for this patch?
  * bulkUpdate
  * createUpdates
  * updateAll

bajtos commented:
> I checked the docs for other CRUD functions, I don't know if any of these apply for this patch?

> * bulkUpdate
> * createUpdates
> * updateAll

These don't apply for this patch.

bajtos commented on lib/dao.js:
```diff
@@ -457,12 +467,27 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
+          if (Model.settings.validateUpsert === undefined) {
+            if (Model.settings.automaticValidation === undefined) {
+
+            } else {
```
Please remove the empty block.

```js
if (Model.settings.automaticValidation !== undefined) {
  doValidate = Model.settings.automaticValidation;
}
```

bajtos commented on lib/dao.js:
```diff
@@ -1876,8 +1911,13 @@ DataAccessObject.prototype.save = function (options, cb) {
+      options.validate = true;
+    } else {
+      options.validate = Model.settings.automaticValidation;
+    }
```
Please rewrite this to use a local variable as we discussed above.

ernie58 replied:
Now that I'm trying: the old code also modified the options-object. When I use a local variable, some tests fail, because they expect options==true when it's undefined

bajtos replied:
I'll still prefer to not modify function arguments (`options`), but since the current code is already modifying it, this patch is not making the matter any worse. Let's leave refactoring to a local variable out of scope of this PR.

bajtos commented on test/optional-validation.test.js:
```diff
@@ -0,0 +1,575 @@
+      if (err) return done(err);
+      d.should.be.instanceOf(User);
+      if (!incompleteData) {
+        d.name.should.eql(OLD_NAME);
```
I find this suspicious. The callback receives an updated model instance, so I would expect the name to equal to the new name? Unless `OLD_NAME` is referring to the value used for `create()` call, in which case a name like `INITIAL_NAME` would be probably better?

bajtos commented:
@ernie58 thank you for the update, you are making a good progress here!

ernie58 commented:
@bajtos I did all the changes, except one (in the save-method) see my comment above.
Could you check again? Thx

ernie58 commented:
Hi @bajtos , could you give an estimate when you could look again at the changes, and when the pull request could be available in the main repository? Just so I can schedule some things. Thx!

bajtos commented:
@slnode ok to test

bajtos commented on test/optional-validation.test.js:
```diff
@@ -0,0 +1,595 @@
+        vip: {type: Boolean}
+      }, { forceId: true, strict: true });
+
+    filter = {fields: ['name', 'id']};
```
AFAICT, this is an unused variable.

bajtos replied:
Please remove.

ernie58 replied:
I've removed it

bajtos commented on test/optional-validation.test.js:
```diff
@@ -0,0 +1,595 @@
+
+    describe('method updateOrCreate with new data', function() {
+      it('should throw on updateOrCreate with validate:true with invalid data', function (done) {
+        User.updateOrCreate(INVALID_DATA, {validate: true}, expectValidationError(done));
```
IIRC, `updateOrCreate` short-circuits to calling `create` when the data does not include any id property. As a result, these tests are checking `create` and not `updateOrCreate`. Could you please check?

ernie58 replied:
That's correct, but the suite of test below does not, i've created 2 describe block, for both scenarios:
  * method updateOrCreate with new data
  * method updateOrCreate with existing data

bajtos replied:
My concern is that the current suite creates a false assumption about what's being tested. We should either remove it or else rewrite to not short-circuit to `create`.

Thoughts?

bajtos replied:
@ernie58 ping ^^^^^

ernie58 replied:
IMHO it's not a false assumption, because it does what is says. It takes new data and uses updateOrCreate. If in the future the short-circuit would be changed to something else, there would be no tests here to check if updateOrCreate uses the optionalValidation in the intended way. Does that makes sense?

But if you think it's better, I'll remove it! Just let me know!

ernie58 replied:
@bajtos what will we decide?

bajtos replied:
@ernie58 I see your point.

Let's remove the whole block `describe('method updateOrCreate with new data')` then, so that we can move this forward.

bajtos commented:
Hi @ernie58, sorry for taking so long to respond. I have one last concern about this patch - see https://github.com/strongloop/loopback-datasource-juggler/pull/752#discussion_r45731801. Other than that, I think this is good to be merged.

ernie58 commented:
@bajtos I've removed the comment and responded to the other issue https://github.com/strongloop/loopback-datasource-juggler/pull/752#discussion_r45734394

ernie58 commented:
@bajtos I've fixed most jshint errors. It would be nice to add a .jshintrc to the repository, that way I could immediately adhere to the code guidelines

bajtos commented:
>  I've fixed most jshint errors. It would be nice to add a .jshintrc to the repository, that way I could immediately adhere to the code guidelines

That's a great idea, would you like to contribute that patch yourself? You can start with jshint & jscs config from loopback repo.

ernie58 commented:
There are quite a lot jshint errors now (more then 600)

ernie58 commented:
@bajtos About the checks failing all the time: is this because something is wrong with the patch I've submitted? If yes, how can I see what has failed so I can make the changes necesary? I'm not that familiar with Jenkins

ernie58 commented:
@bajtos Hi Bajtos, any news on this?

bajtos commented:
> There are quite a lot jshint errors now (more then 600)

Please move these changes to a standalone pull request, so that it's easier to tell in the future which code changes were related to automatic validation and which were purely style fixes.

> About the checks failing all the time: is this because something is wrong with the patch I've submitted? 

The failed checks reported here on GH are caused by downstream repositories. I am afraid our CI build is not exactly stable now, downstream failures are usually unrelated to the changes made :(

Don't worry about them now. Just make sure `npm test` passes on your machine.

> If yes, how can I see what has failed so I can make the changes necesary? I'm not that familiar with Jenkins

I believe our Jenkins logs are not accessible to public yet. We have a parallel Travis CI job that allows contributors to see at least some build results. But as I said before, don't worry about the failed downstream Jenkins job.

bajtos commented:
> There are quite a lot jshint errors now (more then 600)

When we were adding jshint to LoopBack, we have carefully tweaked the config to keep the number of errors at a reasonable level, to avoid too many white-space-only changes. Could you do the same?

ernie58 commented:
@bajtos I've moved the config files to another branch

ernie58 commented:
@bajtos removed the blocks! :)

bajtos commented on lib/dao.js:
```diff
@@ -257,6 +257,16 @@ DataAccessObject.create = function (data, options, cb) {
 
     data = obj.toObject(true);
 
+    //options has precedence on model-setting
```
Please add a space after the slashes.

```
// options etc.
```

bajtos commented on lib/dao.js:
```diff
@@ -257,6 +257,16 @@ DataAccessObject.create = function (data, options, cb) {
+      return create();
+    }
+
+    //only when options.validate is not set, take model-setting into consideration
```
Please add a space after the slashes.

```
// only etc.
```

bajtos commented on lib/dao.js:
```diff
@@ -732,6 +755,16 @@ DataAccessObject.findOrCreate = function findOrCreate(query, data, options, cb)
+          _findOrCreate(query, data, obj);
+        }
+
+        //only when options.validate is not set, take model-setting into consideration
```
Please add a space after the slashes.

```
// only etc.
```

bajtos commented:
@ernie58 thanks. I found three more style details to fix, otherwise the patch LGTM now (031fc34)

Once you have it fixed, could you please squash all commits into a single one and provide a descriptive commit message following the [50/72 rule](tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html)? If you are not familiar with git history rewriting, you are looking for `git rebase -i master` and `git push -f`.

ernie58 commented:
@bajtos done!

bajtos commented:
Landed, thank you for the contribution!

ernie58 commented:
you're welcome
